### PR TITLE
Refactor data generators

### PR DESCRIPTION
**Pull request by florian-huber**
- refactoring of the two main generators used
- they still have a lot of code duplication, but since I am not sure which one will make it I would leave this for later
- added some first tests (currently mostly only checking that the generators return something of right shape)

### CODE REVIEW

File: tests/test_data_generators.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import os
+import json
+import numpy as np
+import pytest
+from matchms import Spectrum
+from ms2deepscore.data_generators import DataGeneratorAllInchikeys
+from ms2deepscore.data_generators import DataGeneratorAllSpectrums
+path_tests  = os.path.dirname(__file__)
```
Comment (svenvanderburg) [inline tests/test_data_generators.py:8]: Should be a constant.

File: tests/test_data_generators.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                                    inchikey_score_mapping, inchikeys_all,
+                                    dim=dimension, batch_size=batch_size,
+                                    num_turns=num_turns, peak_scaling=peak_scaling,
+                                    shuffle=True, ignore_equal_pairs=True,
+                                    same_prob_bins=same_prob_bins,
+                                    augment_peak_removal_max=0.0,
+                                    augment_peak_removal_intensity=0.0,
+                                    augment_intensity=0.0)
+    assert "inchikeys_all and spectrums_binned must have the same dimension." in str(msg.value), \
+        "Expected different expection to be raised"
```
Comment (svenvanderburg) [inline tests/test_data_generators.py:122]:
```suggestion
        "Expected different exception to be raised"
```

File: ms2deepscore/data_generators.py [hunk 3]
```diff
@@ -10,150 +11,188 @@
...
-        listed in *training_ids* num_turns times in every epoch. It will then randomly
-        pick one the spectra corresponding to this InchiKey (if multiple) and pair it
-        with a randomly chosen other spectrum that corresponds to a reference score
-        as defined in same_prob_bins.
 
         Parameters
         ----------
-        spectrums_binned_dicts
+        spectrums_binned
             TODO: Make this less nested and clearer what I should exactly pass.
```
Comment (svenvanderburg) [inline ms2deepscore/data_generators.py:36]: I guess this fixes the TODO?

File: ms2deepscore/data_generators.py [hunk 4]
```diff
@@ -10,150 +11,188 @@
...
+        self.augment_peak_removal_max = augment_peak_removal_max
+        self.augment_peak_removal_intensity = augment_peak_removal_intensity
         self.augment_intensity = augment_intensity
 
     def __len__(self):
         """Denotes the number of batches per epoch"""
         # TODO: this means we don't see all data every epoch, because the last half-empty batch
-        #  is omitted. I guess that is expected behavior?
-        return int(self.num_turns) * int(np.floor(len(self.training_ids) / self.batch_size))
+        #  is omitted. I guess that is expected behavior? --> Yes, with the shuffling in each epoch that seem OK to me (and makes the code easier).
```
Comment (svenvanderburg) [inline ms2deepscore/data_generators.py:105]: Cool :) I guess you can remove the comments now? Or maybe add a small note to the docstring.

File: ms2deepscore/data_generators.py [hunk 5]
```diff
@@ -10,150 +11,188 @@
...
-            TODO: In same order as what?
-            Array of reference similarity scores (=labels).
-        inchikey_mapping
-            TODO: Needs documentation
+        score_array
+            2D-array of reference similarity scores (=labels).
+            Must be symmetric (score_array[i,j] == score_array[j,i]) and in the
+            same order as inchikey_score_mapping.
+        inchikey_ids
+            List of IDs from unique_inchikeys to use for training.
```
Comment (svenvanderburg) [inline ms2deepscore/data_generators.py:43]: `unique_inchickeys` is `inchikey_score_mapping`?

File: tests/test_data_generators.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
+import os
```
Comment (svenvanderburg) [inline tests/test_data_generators.py:1]: Cool! I like these tests 👌  They are simple, but they touch a lot of the code which is what you care about in this stage.

**Review by svenvanderburg — APPROVED** [review]
Really nice changes! Adding the documentation and especially renaming the different ids makes things a lot easier to understand and thus more sustainable! 👍 

There are some PEP8 violations (I spotted a few in the tests), you could add a linter to the CI pipeline, or you could not care :)

File: tests/test_data_generators.py (hunk 6 above)
Comment (florian-huber) [reply tests/test_data_generators.py:1]: Nice, thanks.

File: ms2deepscore/data_generators.py (hunk 3 above)
Comment (florian-huber) [reply ms2deepscore/data_generators.py:36]: Yes you're right I think. Thank.s

File: ms2deepscore/data_generators.py (hunk 4 above)
Comment (florian-huber) [reply ms2deepscore/data_generators.py:105]: Indeed. Will remove that comment.

File: ms2deepscore/data_generators.py (hunk 5 above)
Comment (florian-huber) [reply ms2deepscore/data_generators.py:43]: good catch. thanks!

File: tests/test_data_generators.py (hunk 1 above)
Comment (florian-huber) [reply tests/test_data_generators.py:8]: Sorry, I'll skip that for now...

**Pull request merged**